### PR TITLE
Cancel cold cache prewarming tasks if store is closing

### PR DESCRIPTION
**Pull request by tlrx**
Cold cache prewarming tasks are not stopped immediately when the shard is closed, causing excessive use of disk for nothing. This change adds a `Supplier<Boolean>` to prewarming logic that can be checked before executing any consuming operation to know if the `Store` is closing.

I'm not super happy with my test changes but the logic for checking if prewarming works correctly is tricky.

Closes  #95504

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/cache/full/SearchableSnapshotsPrewarmingIntegTests.java [hunk 1]
```diff
@@ -180,17 +192,22 @@ public void testConcurrentPrewarming() throws Exception {
 
         final Map<String, List<String>> exclusionsPerIndex = new HashMap<>();
         for (int index = 0; index < nbIndices; index++) {
-            exclusionsPerIndex.put("index-" + index, randomSubsetOf(Arrays.asList("fdt", "fdx", "nvd", "dvd", "tip", "cfs", "dim")));
+            exclusionsPerIndex.put(
+                "index-" + index,
+                randomSubsetOf(Stream.of(LuceneFilesExtensions.values()).map(LuceneFilesExtensions::getExtension).toList())
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/cache/full/SearchableSnapshotsPrewarmingIntegTests.java:197]: A small improvement

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/cache/full/SearchableSnapshotsPrewarmingIntegTests.java [hunk 2]
```diff
@@ -236,18 +253,72 @@ public void testConcurrentPrewarming() throws Exception {
...
+            assertBusy(() -> deletedIndices.forEach(deletedIndex -> assertThat(indicesService.hasIndex(deletedIndex), is(false))));
+        }
+
+        startPrewarmingLatch.countDown();
+
+        // wait for recovery to be DONE
         assertBusy(() -> {
-            final ThreadPool threadPool = getInstanceFromNode(ThreadPool.class);
-            assertThat(threadPool.info(SearchableSnapshots.CACHE_FETCH_ASYNC_THREAD_POOL_NAME).getQueueSize(), nullValue());
-            assertThat(threadPool.info(SearchableSnapshots.CACHE_PREWARMING_THREAD_POOL_NAME).getQueueSize(), nullValue());
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/cache/full/SearchableSnapshotsPrewarmingIntegTests.java:250]: This was wrong

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/CachedBlobContainerIndexInput.java [hunk 3]
```diff
@@ -129,19 +129,26 @@ protected void readWithoutBlobCache(ByteBuffer b) throws Exception {
...
-     * @return a tuple with {@code Tuple<Persistent Cache Length, Prefetched Length>} values
+     *
+     * @param part the index of the part to prewarm
+     * @param isCancelled a {@link Supplier<Boolean>} that allows to check if prewarming must be cancelled
+     *
+     * @return the number of bytes that has been read from the blob store when prewarming the part,
+     * or {@code -1} if the prewarming was cancelled
      */
-    public Tuple<Long, Long> prefetchPart(final int part) throws IOException {
+    public long prefetchPart(final int part, Supplier<Boolean> isCancelled) throws IOException {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/CachedBlobContainerIndexInput.java:139]: Only one part of the tuple was used

**Review by original-brownbear — APPROVED** [review]
LGTM Thanks for handling this Tanguy!

**Comment by tlrx** [comment]
Thanks Armin!

**Comment by tlrx** [comment]
I opened #96024 to backport to 7.17.11, with small adjustments since `RefCountingListener` and `ThrottledTaskRunner` are not available there.

**Pull request merged**